File: objective-c-sample/SendBirdSampleUI/SendBird/SendBirdMessaging/View/MessagingMyMessageTableViewCell.m
//
//  MessagingMyMessageTableViewCell.m
//  SendBirdSampleUI
//
//  Created by SendBird Developers on 2015. 7. 29..
//  Copyright (c) 2015 SENDBIRD.COM. All rights reserved.
//

#import "MessagingMyMessageTableViewCell.h"

#define kMyMessageCellTopMargin 14
#define kMyMessageCellBottomMargin 0
#define kMyMessageCellLeftMargin 12
#define kMyMessageBalloonRightMargin 12
#define kMyMessageCellRightMargin 32
#define kMyMessageFontSize 14.0
#define kMyMessageBalloonTopPadding 12
#define kMyMessageBalloonBottomPadding 12
#define kMyMessageBalloonLeftPadding 12
#define kMyMessageBalloonRightPadding 12
#define kMyMessageMaxWidth 168
#define kMyMessageDateTimeRightMarign 4
#define kMyMessageDateTimeFontSize 10.0
#define kMyMessageUnreadFontSize 10.0

@implementation MessagingMyMessageTableViewCell {
    CGFloat topMargin;
}

- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    
    if (self != nil) {
        topMargin = kMyMessageCellTopMargin;
        [self initViews];
    }
    
    return self;
}

- (void) initViews
{
    [self setBackgroundColor:[UIColor clearColor]];
    
    self.messageBackgroundImageView = [[UIImageView alloc] init];
    [self.messageBackgroundImageView setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.messageBackgroundImageView setImage:[UIImage imageNamed:@"_bg_chat_bubble_purple"]];
    [self addSubview:self.messageBackgroundImageView];
    
    self.messageLabel = [[UILabel alloc] init];
    [self.messageLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.messageLabel setFont:[UIFont systemFontOfSize:14]];
    [self.messageLabel setNumberOfLines:0];
    [self.messageLabel setLineBreakMode:NSLineBreakByCharWrapping];
    [self addSubview:self.messageLabel];
    
    self.dateTimeLabel = [[UILabel alloc] init];
    [self.dateTimeLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.dateTimeLabel setNumberOfLines:1];
    [self.dateTimeLabel setTextColor:UIColorFromRGB(0xacaab2)];
    [self.dateTimeLabel setFont:[UIFont systemFontOfSize:kMyMessageDateTimeFontSize]];
    [self.dateTimeLabel setText:@"11:24 PM"];
    [self addSubview:self.dateTimeLabel];
    
    self.unreadLabel = [[UILabel alloc] init];
    [self.unreadLabel setTranslatesAutoresizingMaskIntoConstraints:NO];
    [self.unreadLabel setNumberOfLines:1];
    [self.unreadLabel setTextColor:UIColorFromRGB(0xac90ff)];
    [self.unreadLabel setFont:[UIFont systemFontOfSize:kMyMessageUnreadFontSize]];
    [self.unreadLabel setText:@"Unread"];
    [self.unreadLabel setHidden:YES];
    [self addSubview:self.unreadLabel];
    
    // Message Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageLabel
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:-kMyMessageBalloonBottomPadding]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-kMyMessageCellRightMargin]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageLabel
                                                     attribute:NSLayoutAttributeWidth
                                                     relatedBy:NSLayoutRelationLessThanOrEqual
                                                        toItem:nil
                                                     attribute:NSLayoutAttributeNotAnAttribute
                                                    multiplier:1 constant:kMyMessageMaxWidth]];
    
    // Message Background Image View
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:-kMyMessageCellBottomMargin]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeLeading
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageLabel
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:-kMyMessageBalloonLeftPadding]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self
                                                     attribute:NSLayoutAttributeTrailing
                                                    multiplier:1 constant:-kMyMessageBalloonRightMargin]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeTop
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageLabel
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:-kMyMessageBalloonTopPadding]];
    
    // DateTime Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.dateTimeLabel
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeBottom
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.dateTimeLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:-kMyMessageDateTimeRightMarign]];
    
    // Unread Label
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadLabel
                                                     attribute:NSLayoutAttributeBottom
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.dateTimeLabel
                                                     attribute:NSLayoutAttributeTop
                                                    multiplier:1 constant:0]];
    [self addConstraint:[NSLayoutConstraint constraintWithItem:self.unreadLabel
                                                     attribute:NSLayoutAttributeTrailing
                                                     relatedBy:NSLayoutRelationEqual
                                                        toItem:self.messageBackgroundImageView
                                                     attribute:NSLayoutAttributeLeading
                                                    multiplier:1 constant:-kMyMessageDateTimeRightMarign]];
}

- (void) setContinuousMessage:(BOOL)continuousFlag
{
    if (continuousFlag) {
        topMargin = 4.0;
    }
    else {
        topMargin = kMyMessageCellTopMargin;
    }
}

- (void) setModel:(SendBirdMessage *)message
{
    self.message = message;
    [self.messageLabel setAttributedText:[self buildMessage]];
    long long ts = [self.message getMessageTimestamp] / 1000;
    [self.dateTimeLabel setText:[SendBirdUtils messageDateTime:ts]];
    [self.unreadLabel setHidden:YES];
    
    int unreadCount = 0;
    if (self.readStatus != nil) {
        for (NSString *key in self.readStatus) {
            if (![key isEqualToString:[SendBird getUserId]]) {
                long long readTime = [[self.readStatus objectForKey:key] longLongValue] / 1000;
                if (ts <= readTime) {
                    //                    [self.unreadLabel setHidden:YES];
                }
                else {
                    unreadCount = unreadCount + 1;
                }
            }
        }
    }
    
    if (unreadCount == 0) {
        [self.unreadLabel setHidden:YES];
    }
    else {
        [self.unreadLabel setHidden:NO];
        [self.unreadLabel setText:[NSString stringWithFormat:@"Unread %d", unreadCount]];
    }
}

- (NSAttributedString *)buildMessage
{
    NSMutableDictionary *messageAttribute = [NSMutableDictionary dictionaryWithObjectsAndKeys:[UIFont systemFontOfSize:kMyMessageFontSize], NSFontAttributeName, UIColorFromRGB(0x3d3d3d), NSForegroundColorAttributeName,nil];
    NSMutableDictionary *urlAttribute = [NSMutableDictionary dictionaryWithObjectsAndKeys:[UIFont systemFontOfSize:kMyMessageFontSize], NSFontAttributeName, UIColorFromRGB(0x2981e1), NSForegroundColorAttributeName,nil];
    
    NSString *message = [[NSString stringWithFormat:@"%@", [self.message message]] stringByReplacingOccurrencesOfString:@" " withString:@"\u00A0"];
    NSString *url = [SendBirdUtils getUrlFromString:[self.message message]];
    NSRange urlRange;
    if ([url length] > 0) {
        urlRange = [message rangeOfString:url];
    }
    message = [message stringByReplacingOccurrencesOfString:@" " withString:@"\u00A0"];
    message = [message stringByReplacingOccurrencesOfString:@"-" withString:@"\u2011"];
    
    NSMutableAttributedString *attributedMessage = [[NSMutableAttributedString alloc] initWithString:message];
    NSRange messageRange = NSMakeRange(0, [[self.message message] length]);
    
    [attributedMessage beginEditing];
    [attributedMessage setAttributes:messageAttribute range:messageRange];
    if ([url length] > 0) {
        [attributedMessage setAttributes:urlAttribute range:urlRange];
    }
    [attributedMessage endEditing];
    
    return attributedMessage;
}

- (CGFloat)getHeightOfViewCell:(CGFloat)totalWidth
{
    CGRect messageRect;
    NSAttributedString *attributedMessage = [self buildMessage];
    
    messageRect = [attributedMessage boundingRectWithSize:CGSizeMake(kMyMessageMaxWidth, CGFLOAT_MAX) options:(NSStringDrawingUsesLineFragmentOrigin) context:nil];
    
    return messageRect.size.height + topMargin + kMyMessageCellBottomMargin + kMyMessageBalloonTopPadding + kMyMessageBalloonBottomPadding;
}

@end

File: swift-sample/SendBirdSampleUI/SendBird/SendBirdMessaging/View/MessagingMyMessageTableViewCell.swift
//
//  MessagingMyMessageTableViewCell.swift
//  SendBirdSampleUI
//
//  Created by Jed Kyung on 2/7/16.
//  Copyright © 2016 SENDBIRD.COM. All rights reserved.
//

import UIKit
import SendBirdSDK

class MessagingMyMessageTableViewCell: UITableViewCell {
    let kMyMessageCellTopMargin: CGFloat = 14
    let kMyMessageCellBottomMargin: CGFloat = 0
    let kMyMessageCellLeftMargin: CGFloat = 12
    let kMyMessageBalloonRightMargin: CGFloat = 12
    let kMyMessageCellRightMargin: CGFloat = 32
    let kMyMessageFontSize: CGFloat = 14.0
    let kMyMessageBalloonTopPadding: CGFloat = 12
    let kMyMessageBalloonBottomPadding: CGFloat = 12
    let kMyMessageBalloonLeftPadding: CGFloat = 12
    let kMyMessageBalloonRightPadding: CGFloat = 12
    let kMyMessageMaxWidth: CGFloat = 168
    let kMyMessageDateTimeRightMarign: CGFloat = 4
    let kMyMessageDateTimeFontSize: CGFloat = 10.0
    let kMyMessageUnreadFontSize: CGFloat = 10.0
    
    var message: SendBirdMessage?
    var messageBackgroundImageView: UIImageView?
    var messageLabel: UILabel?
    var dateTimeLabel: UILabel?
    var unreadLabel: UILabel?
    var readStatus: NSMutableDictionary?
    
    private var topMargin: CGFloat?

    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        self.topMargin = kMyMessageCellTopMargin
        self.initViews()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func initViews() {
        self.backgroundColor = UIColor.clearColor()
        
        self.messageBackgroundImageView = UIImageView()
        self.messageBackgroundImageView?.translatesAutoresizingMaskIntoConstraints = false
        self.messageBackgroundImageView?.image = UIImage.init(named: "_bg_chat_bubble_purple")
        self.addSubview(self.messageBackgroundImageView!)
        
        self.messageLabel = UILabel()
        self.messageLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.messageLabel?.font = UIFont.systemFontOfSize(14.0)
        self.messageLabel?.numberOfLines = 0
        self.messageLabel?.lineBreakMode = NSLineBreakMode.ByCharWrapping
        self.addSubview(self.messageLabel!)
        
        self.dateTimeLabel = UILabel()
        self.dateTimeLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.dateTimeLabel?.numberOfLines = 1
        self.dateTimeLabel?.textColor = SendBirdUtils.UIColorFromRGB(0xacaab2)
        self.dateTimeLabel?.font = UIFont.systemFontOfSize(kMyMessageDateTimeFontSize)
        self.dateTimeLabel?.text = "11:24 PM"
        self.addSubview(self.dateTimeLabel!)
        
        self.unreadLabel = UILabel()
        self.unreadLabel?.translatesAutoresizingMaskIntoConstraints = false
        self.unreadLabel?.numberOfLines = 1
        self.unreadLabel?.textColor = SendBirdUtils.UIColorFromRGB(0xac90ff)
        self.unreadLabel?.font = UIFont.systemFontOfSize(kMyMessageUnreadFontSize)
        self.unreadLabel?.text = "Unread"
        self.unreadLabel?.hidden = true
        self.addSubview(self.unreadLabel!)
        
        // Message Label
        self.addConstraint(NSLayoutConstraint.init(item: self.messageLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: -kMyMessageBalloonBottomPadding))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -kMyMessageCellRightMargin))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageLabel!, attribute: NSLayoutAttribute.Width, relatedBy: NSLayoutRelation.LessThanOrEqual, toItem: nil, attribute: NSLayoutAttribute.NotAnAttribute, multiplier: 1, constant: kMyMessageMaxWidth))
        
        // Message Background Image View
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: -kMyMessageCellBottomMargin))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Leading, relatedBy: NSLayoutRelation.Equal, toItem: self.messageLabel!, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: -kMyMessageBalloonLeftPadding))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self, attribute: NSLayoutAttribute.Trailing, multiplier: 1, constant: -kMyMessageBalloonRightMargin))
        self.addConstraint(NSLayoutConstraint.init(item: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Top, relatedBy: NSLayoutRelation.Equal, toItem: self.messageLabel!, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: -kMyMessageBalloonTopPadding))
        
        // DateTime Label
        self.addConstraint(NSLayoutConstraint.init(item: self.dateTimeLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Bottom, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.dateTimeLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: -kMyMessageDateTimeRightMarign))
        
        // Unread Label
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadLabel!, attribute: NSLayoutAttribute.Bottom, relatedBy: NSLayoutRelation.Equal, toItem: self.dateTimeLabel!, attribute: NSLayoutAttribute.Top, multiplier: 1, constant: 0))
        self.addConstraint(NSLayoutConstraint.init(item: self.unreadLabel!, attribute: NSLayoutAttribute.Trailing, relatedBy: NSLayoutRelation.Equal, toItem: self.messageBackgroundImageView!, attribute: NSLayoutAttribute.Leading, multiplier: 1, constant: -kMyMessageDateTimeRightMarign))
    }

    func setContinuousMessage(continuousFlag: Bool) {
        if continuousFlag {
            self.topMargin = 4.0
        }
        else {
            self.topMargin = kMyMessageCellTopMargin
        }
    }
    
    func setModel(message: SendBirdMessage) {
        self.message = message
        self.messageLabel?.attributedText = self.buildMessage()
        let ts: Int64 = (self.message?.getMessageTimestamp())! / 1000
        self.dateTimeLabel?.text = SendBirdUtils.messageDateTime(NSTimeInterval(ts))
        self.unreadLabel?.hidden = true
        
        var unreadCount: Int = 0
        if self.readStatus != nil {
            for item in self.readStatus! {
                if item.key as? String != SendBird.getUserId() {
                    let readTime: Int64 = (item.value as! NSNumber).longLongValue 
                    if ts <= readTime {
                        
                    }
                    else {
                        unreadCount = unreadCount + 1
                    }
                }
            }
        }
        
        if unreadCount == 0 {
            self.unreadLabel?.hidden = true
        }
        else {
            self.unreadLabel?.hidden = false
            self.unreadLabel?.text = String.init(format: "Unread %d", unreadCount)
        }
    }
    
    func buildMessage() -> NSAttributedString {
        let messageAttribute = [NSFontAttributeName: UIFont.systemFontOfSize(kMyMessageFontSize), NSForegroundColorAttributeName: SendBirdUtils.UIColorFromRGB(0x3d3d3d)]
        let urlAttribute = [NSFontAttributeName: UIFont.systemFontOfSize(kMyMessageFontSize), NSForegroundColorAttributeName: SendBirdUtils.UIColorFromRGB(0x2981e1)]
        
        var message: NSString = NSString.init(format: "%@", (self.message?.message)!)
        let url: String = SendBirdUtils.getUrlFromstring((self.message?.message)!)
        var urlRange: NSRange?
        if url.characters.count > 0 {
            urlRange = message.rangeOfString(url)
        }
        message = message.stringByReplacingOccurrencesOfString(" ", withString: "\u{00A0}")
        message = message.stringByReplacingOccurrencesOfString(" ", withString: "\u{2011}")
        
        let attributedMessage: NSMutableAttributedString = NSMutableAttributedString.init(string: message as String)
        let messageRange: NSRange = NSMakeRange(0, self.message!.message.characters.count)
        
        attributedMessage.beginEditing()
        attributedMessage.setAttributes(messageAttribute, range: messageRange)
        if url.characters.count > 0 {
            attributedMessage.setAttributes(urlAttribute, range: urlRange!)
        }
        attributedMessage.endEditing()
        
        return attributedMessage
    }
    
    func getHeightOfViewCell(totalWidth: CGFloat) -> CGFloat {
        var messageRect: CGRect
        let attributedMessage: NSAttributedString = self.buildMessage()
        
        messageRect = attributedMessage.boundingRectWithSize(CGSizeMake(kMyMessageMaxWidth, CGFloat.max), options: NSStringDrawingOptions.UsesLineFragmentOrigin, context: nil)
        
        return messageRect.size.height + self.topMargin! + kMyMessageCellBottomMargin + kMyMessageBalloonTopPadding + kMyMessageBalloonBottomPadding
    }
}
